core/base/range.hpp

```diff
@@ -0,0 +1,966 @@
+Copyright 2017-2018
+
+Karlsruhe Institute of Technology
+Universitat Jaume I
+University of Tennessee
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_BASE_RANGE_HPP_
+#define GKO_CORE_BASE_RANGE_HPP_
+
+
+#include "core/base/math.hpp"
+#include "core/base/std_extensions.hpp"
+#include "core/base/types.hpp"
+#include "core/base/utils.hpp"
+
+
+namespace gko {
+
+
+/**
+ * A span is a lightweight structure used to create sub-ranges from other
+ * ranges.
+ *
+ * A `span s` represents a contiguous set of indexes in one dimension of the
+ * range, starting on index `s.begin` (inclusive) and ending at index `s.end`
+ * (exclusive). A span is only valid if its starting index is not greater than
+ * its ending index. If `s.begin == s.end`, span `s` represents an empty set of
+ * indexes.
+ *
+ * Spans can be compared using the `==` and `!=` operators. Two spans are
+ * identical if both their `begin` and `end` values are identical. Note that
+ * this implies there are "empty" spans which are not considered identical.
+ *
+ * Spans also have two distinct partial orderings defined on them:
+ * 1.  `x < y` (`y > x`) if and only if `x.end < y.begin`
+ * 2.  `x <= y` (`y >= x`) if and only if `x.end <= y.begin`
+ *
+ * Note that the ordering is in fact partial - there are spans `x` and `y` for
+ * which none of the following inequalities holds:
+ *     `x < y`, `x > y`, `x == y`, `x <= y`, `x >= y`.
```

Comment:
Shouldn't atleast one inequality hold ?

Reply:
Do you mean because of antisymmetry?

One thing that can be a bit confusing is that I use `<=` and `<` as two different relations, not as the inclusive/exclusive version of the same relation. (Formally, `<=` &ne; `<` &cup; `==`). Also, `==` &#8840; `<=`. Unfortunately, C++ has a very limited set of operators, so we have to make due with what we have.

Both  `<` and `<=` are _exclusive_ (strict) relations, and you can define _inclusive_ (weak) versions __R1__ &equiv; (`<` &cup; `==`), and __R2__ &equiv; (`<=` &cup; `==`) of both of them. The antisymmetry properties (x __R1__ y) &and; (y __R1__ x) &rArr; (x `==` y) and (x __R2__ y) &and; (y __R2__ x) &rArr; (x `==` y) hold for both versions.

---------

I did find a small problem with `<=` though. I claimed that this is a strict partial order, which menas that the following should hold:

1. not x `<=` x  (irreflexivity)
2. if a `<=` b and b `<=` c then a `<=` c (transitivity)

However, for empty spans, irreflexivity does not hold (`[t, t) <= [t,t)` is true).

We could fix it and make it a true strict partial order by defining `[x, y) <= [z, w)` &hArr; y &le; z &and; &not; (x = y = z = w), but that just reduces performance to support a trivial use-case which will probably never be needed in practice.

Reply:
No, I mean here  `span s` is a subset of a `range r`, so if you take a `span x` and a `span y` &#8838; `r` then shouldn't atleast one of `==`, `<=`, `>=` ,`<`, `>` be satisfied ? For example consider two spans `x` and `y` in `r`, then if none of the properties are satisfied then that would imply that they are not in the same range `r`, because any two sub-sequences belonging to a larger sequence should be closed by the above relations.

Reply:
`<=` and `<` are not subset relations

Reply:
Actually, I'm not sure what you're trying to say. Span's don't even belong to a range, they are completely independent.

Reply:
I now see I must have deleted the example I had before submitting my original response: `x = [2, 4)` and `y = [3, 5)` is an example when none of the relations holds.

Reply:
Sorry, I misunderstood what you were saying. I get it now. Even though the strict partial order is a bit confusing, but I understand the true strict partial ordering can reduce performance and I am okay with how it is now. I see that you have mentioned this exception in the documentation as well, so that is good.